wdl-lint/src/rules/inconsistent_newlines.rs

```diff
@@ -0,0 +1,83 @@
+//! A lint rule for ensuring that newlines are consistent.
+
+use wdl_ast::AstNode;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Span;
+use wdl_ast::SyntaxKind;
+use wdl_ast::ToSpan;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the inconsistent newlines rule.
+const ID: &str = "InconsistentNewlines";
+
+/// Creates an inconsistent newlines diagnostic.
+fn inconsistent_newlines(span: Span) -> Diagnostic {
+    Diagnostic::note("inconsistent newlines detected")
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_fix("use either \"\\n\" or \"\\r\\n\" consistently in the file")
+}
+
+/// Detects imports that are not sorted lexicographically.
+#[derive(Debug, Clone, Copy)]
+pub struct InconsistentNewlinesRule;
+
+impl Rule for InconsistentNewlinesRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that newline usage is consistent."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Files should not mix \\n and \\r\\n line breaks. Pick one and use it consistently in your \
+         project."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn visitor(&self) -> Box<dyn Visitor<State = Diagnostics>> {
+        Box::new(InconsistentNewlinesVisitor)
+    }
+}
+
+/// Implements the visitor for the import sort rule.
+struct InconsistentNewlinesVisitor;
+
+impl Visitor for InconsistentNewlinesVisitor {
+    type State = Diagnostics;
+
+    fn document(&mut self, state: &mut Self::State, reason: VisitReason, doc: &wdl_ast::Document) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        let mut newline = 0;
+        let mut carriage_return = 0;
+
+        doc.syntax()
```

Comment:
So this walks only the children of the root, not all the whitespace tokens in the document.

I would suggest implementing the `whitespace` trait method to do the actual detection, setting fields on the visitor to denote if we've detected `\r\n` and `\n`, then in the `document` trait method add the diagnostic on exit.

The `whitespace` method will be called for every whitespace token in the file, regardless of where they appear.

Reply:
I've done this, but when is a token not a child of the document root?

Reply:
Whitespace tokens are only (immediate) children of the root if the whitespace appears outside other node children of the root.

For example:

```wdl
version 1.1

# The above and below whitespace are children of the root and siblings to the workflow definition node.

workflow foo {

    # The whitespace in here is a child of the workflow definition node, however.

}
```

While `descendants_with_tokens` could have been used instead to process each descendant whitespace token, that would amount to another full walk of the AST, which the validator is doing already when calling each visitor.

Reply:
Ah, that's it `children_with_tokens()` is the _immediate_ children of the `root` not a recursive set of children.